autogo/shadow.go

```diff
@@ -0,0 +1,158 @@
+func shadowClone(origin, dest string) error {
+	v := visitor{
+		origin: origin,
+		dest:   dest,
+	}
+	return filepath.Walk(v.origin, v.visit)
+}
+
+// action does something with the current file (mkdir or link)
+type action func(origin string, info os.FileInfo, dest string) error
+
+// visitor stores the origin we're cloning from and destionation we clone to.
+type visitor struct {
+	// origin is the path we read to determine what to clone
+	origin string
+	// dest is where we clone
+	dest string
+}
+
+// visit chooses and then performs the right action for the current path
+func (v visitor) visit(path string, info os.FileInfo, verr error) error {
+	act, dest, err := v.choose(path, info, verr)
+	if err != nil {
+		return err
+	}
+	if act == nil {
+		return nil
+	}
+	return act(path, info, dest)
+}
+
+// choose looks at the current path and returns the appropriate action:
+// - testdata directory => skip it and everything under it
+// - directory => create the directory
+// - go file (.go, .s) => create it at dest
+// - other files => do nothing
+func (v visitor) choose(path string, info os.FileInfo, verr error) (action, string, error) {
+	if verr != nil {
+		return nil, "", fmt.Errorf("failed to walk to %s: %v", path, verr)
+	}
+
+	// If origin is /foo/bar and we receive /foo/bar/something/special
+	// Change this to something/special
+	r, err := filepath.Rel(v.origin, path)
+	// First ensure the path is a child of origin
+	if err != nil {
+		return nil, "", fmt.Errorf("%q is not relative to %q: %v", path, v.origin, err)
+	}
+	if r == ".." || strings.HasPrefix(r, "../") {
+		return nil, "", fmt.Errorf("%q is not a child of %q", path, v.origin)
+	}
+
+	// TODO(fejta): handle symlink loops better
+	// Many repos (dep is one) have testdata folders which are
+	// either symlink loops back to some other part of the repo
+	// or else intentionally do not compile.
+	//
+	// The gazelle tool is not robust to packages containing files that fail to compile.
+	//
+	// For now just ignore these testdata folders (usually in vendored packages). This makes bazel build work at the cost of breaking bazel test
+	if info.IsDir() && strings.HasSuffix(path, "/testdata") {
+		log.Printf("Skipping %s...", path)
+		return nil, "", filepath.SkipDir
+	}
+
+	d := filepath.Join(v.dest, r)
+	// Create dirs
+	if info.IsDir() {
+		return mkdir, d, nil
+	}
+
+	if strings.Contains(path, "/testdata/") {
+		return nil, "", fmt.Errorf("%s is within a testdata dir, which should not happen", path)
+	}
+
+	// Ignore files irrelevant to go
+	if !strings.HasSuffix(path, ".go") && !strings.HasSuffix(path, ".s") {
+		// Assume .go and .s are the only relevant files to a build
+		// TODO(fejta): consider adding BUILD and BUILD.bazel files, or at least the go rules within them
+		return nil, "", nil
+	}
+
+	// Link in golang files
+	return link, d, nil
+}
+
+// mkdir creates directories for dest
+var mkdir = func(_ string, info os.FileInfo, dest string) error {
```

Comment:
why isn't this `func mkdir()` (same for link)?

Reply:
Done

This was a failed attempt to compare function values in the unit test, which is [unsupported](https://golang.org/pkg/reflect/#DeepEqual): `Func values are deeply equal if both are nil; otherwise they are not deeply equal.`